src/scene/materials/material.js

```diff
@@ -380,10 +381,39 @@ class Material {
     updateUniforms(device, scene) {
     }
 
-    getShaderVariant(device, scene, objDefs, staticLightList, pass, sortedLights) {
-        // return the shader specified by the user of the material
-        Debug.assert(this._shader, 'Material does not have shader set', this);
-        return this._shader;
+    getShaderVariant(device, scene, objDefs, staticLightList, pass, sortedLights, viewUniformFormat, viewBindGroupFormat) {
+
+        Debug.assert(this._shader, 'Material does not have a shader set', this);
+
+        // default generator for a material - simply return existing shader definition. Use generator and getProgram
+        // to allow for shader processing to be cached
+        const key = `shader-id-${this._shader.id}`;
+        const shaderDefinition = this._shader.definition;
+        const materialGenerator = {
+            generateKey: function (options) {
+                // unique name based of the shader id
+                return key;
+            },
+
+            createShaderDefinition: function (device, options) {
+                return shaderDefinition;
+            }
+        };
+
+        // temporarily register the program generator
+        const libraryModuleName = 'shader';
+        const library = device.getProgramLibrary();
+        Debug.assert(!library.isRegistered(libraryModuleName));
+        library.register(libraryModuleName, materialGenerator);
```

Comment:
having to register then deregister is pretty messy. can the code be reshuffled to expose getProgram on the generator directly?